WIP: Cython port of contact map functions

kain88-de opened the pull request:
I want to fix #264. The non PBC version is already implemented in Cython and the tests are passing as well. The Cython-implementation is also faster using the 1Q0W protein.

Porting the PBC version I noticed that there are no tests for it. I added a test and the test suite is still passing. But I'm not sure if the test I have added is run. Btw I'm new to nose is there a way to specify a subset of tests that I want to run instead of executing the whole test-suite all the time?

If someone can help me get the test-cases for the PBC-case running I will also port the PBC version.

---------

Here is a short description how I run the test-suite since I couldn't find up-to-date information how to set-up a dev environment or run unit-tests.

For testing I'm using the approach defined in the `travis.yml`, but I installed all dependencies via pip.
After each change in either the package or testsuite folder I reinstalled them with `pip install --upgrade <folder>/`

richardjgowers commented:
Hey,

This looks like a good start.

To run a subset of tests, from a command line in the directory `testsuite/MDAnalysisTests` you can do 
```
# run tests from a specific file
nosetests test_analysis.py

# or specify the name of the class to run
nosetests test_analysis.py:TestContactMatrix
```
More details here:
https://github.com/MDAnalysis/mdanalysis/wiki/UnitTests

This should then let you see the two tests that fail on Travis (scroll to bottom):
https://travis-ci.org/MDAnalysis/mdanalysis/builds/73811170

If you're not sure a test is running, the quickest way to check is to change the reference state (your `res`) and see if the test then fails.

Let us know if you need any more help

kain88-de commented:
Ok I got the new test running and it passes like it should.

The current sparse test is failing though due to an error in the cython code. My next problem is that installing the changes with `pip install --upgrade packages/` isn't updating the cython-generated C-module. Is there a special way how I should trigger the cython-compiler?

Sorry for all the questions but this is badly/not documented in the wiki.

richardjgowers commented:
My workflow is to run `setup.py install --user` from the `package/` directory, I'm not sure if this is the same as your method.

If that doesn't do it, deleting `src/numtools/distances.c` should force it to regenerate the C module from the pyx you wrote.

kain88-de commented:
Yeah got it working now. Installing cython in the virtual-env and removing the c file did it.

I have noticed that the return values between the numpy and sparse matrix are inconsistent. For the numpy array the diagonal of the contact-matrix is true. But for sparse matrices the diagonal is set to false.  

Personally I would prefer if if the diagonal elements were always true independent of the chosen return-type.

richardjgowers commented:
It's not documented either way, so we can go with the diagonal being True (update the docstring and tests) unless anyone objects.

The only thing in the package that uses contact_matrix current is `analysis/leaflet`, and I think all this does is build a networkx graph from the contacts, I'm not sure how the diagonal would affect this? @orbeckst 

To do the PBC versions, it might be a good idea to import `minimum_image` from `calc_distances.h` and use that.  I think you'll need to change your `x` `y` `z` into a single array for that function.

kain88-de commented:
Btw is the progressbar feature still required? Without it it should be straight forward to add these functions to `calc_distances.h` which will make adding openmp support later easier.

richardjgowers commented:
I wasn't too sure how easy it'd be to fill a np sparse matrix in C though?

orbeckst commented:
The progressbar feature is not really necessary inside the cython code; feel free to get rid of it, especially if it makes the code cleaner.

kain88-de commented:
I added some changes

- removed progress bar
- only calculate upper triangle matrix and use symmetry
- refactored test to be nicer to read and easier to understand.
- set diagonal elements to `True` also for sparse matrices

richardjgowers commented:
Ok this is looking good, nice catch with the reflection.

There's actually a minor bug in the code you've translated, (was present before so not your fault), I've left a line comment, if you can fix that up then we're good to go.

Other minor things to do:
 - add yourself to `AUTHORS`
 - add a line to `CHANGELOG` detailing the fix to the diagonal return type

richardjgowers commented on package/src/numtools/distances.pyx:
```diff
@@ -255,3 +256,56 @@ def triclinic_pbc(c_numpy.ndarray coords,
+            y = xyz[i, 1] - xyz[j, 1]
+            z = xyz[i, 2] - xyz[j, 2]
+
+            if abs(x) > box_half[0]:
```
This will only add/subtract a single `box_view` from each dimension, but it needs to do it multiple times where necessary.

So if `box_half = 10` then `x  = 1 = 11 == 21 = 31`, and `x = 1 = -9 = -19`

You can either import the `minimum_image` function from `calc_distances.h` which does this, or inline that into here.

kain88-de replied:
Good catch.  I always assume that particles are in one of the neighboring images and then wonder that happened when this is not the case.  I will first add this to the unit-tests and then fix it.

kain88-de commented:
PBC are handled correctly with `minium-image` now.

kain88-de commented:
Btw should I regenerate the documentation or is this done automatically with a git-hook when this PR is merged?

orbeckst commented:
Docs are not generated automatically (yet). There is a page in the wiki outlining the procedure but in the end someone with write access has to merge the develop branch with docs into the gh-pages branch. 

--
Oliver Beckstein
email: orbeckst@gmail.com
skype: orbeckst


Am Aug 4, 2015 um 2:55 schrieb kain88-de <notifications@github.com>:

> Btw should I regenerate the documentation or is this done automatically with a git-hook when this PR is merged?
> 
> —
> Reply to this email directly or view it on GitHub.
>

kain88-de commented:
Docs are updated now. Sphinx has rebuild almost all the doc pages but I have only included the one affected by this PR.

orbeckst commented:
I had a quick look and it seems fine but I let @richardjgowers do the final approval and merge.

Thanks @kain88-de !
